nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/status/analytics/StatusAnalyticsModelMapFactory.java

```diff
@@ -0,0 +1,126 @@
+     * @param nifiProperties NiFi Properties object
+     * @return statusAnalyticsModel
+     */
+    private static StatusAnalyticsModel createModelInstance(ExtensionManager extensionManager, NiFiProperties nifiProperties) {
+        final String implementationClassName = nifiProperties.getProperty(NiFiProperties.ANALYTICS_CONNECTION_MODEL_IMPLEMENTATION, NiFiProperties.DEFAULT_ANALYTICS_CONNECTION_MODEL_IMPLEMENTATION);
+        if (implementationClassName == null) {
+            throw new RuntimeException("Cannot create Analytics Model because the NiFi Properties is missing the following property: "
+                    + NiFiProperties.ANALYTICS_CONNECTION_MODEL_IMPLEMENTATION);
+        }
+        try {
+            return NarThreadContextClassLoader.createInstance(extensionManager, implementationClassName, StatusAnalyticsModel.class, nifiProperties);
+        } catch (final Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Get a connection status extract function instance
+     * @return StatusMetricExtractFunction
+     */
+    private static StatusMetricExtractFunction getConnectionStatusExtractFunction() {
+
+        return (metric, statusHistory) -> {
+
+            List<Double> values = new ArrayList<>();
+            List<Double[]> features = new ArrayList<>();
+            Random rand = new Random();
+            StatusHistoryDTO statusHistoryDTO = StatusHistoryUtil.createStatusHistoryDTO(statusHistory);
+
+            for (StatusSnapshotDTO snap : statusHistoryDTO.getAggregateSnapshots()) {
+                List<Double> featureArray = new ArrayList<>();
+                Long snapValue = snap.getStatusMetrics().get(metric);
+                long snapTime = snap.getTimestamp().getTime();
+
+                featureArray.add((double) snapTime);
+                Double randomError = +(rand.nextInt(1000) * .0000001);
+                if (metric.equals(QUEUED_COUNT_METRIC)) {
+
+                    Long inputCount = snap.getStatusMetrics().get(INPUT_COUNT_METRIC);
+                    Long outputCount = snap.getStatusMetrics().get(OUTPUT_COUNT_METRIC);
+                    Double inOutRatio = ((double) outputCount / (double) inputCount) + randomError;
+                    featureArray.add(Double.isNaN(inOutRatio) ? randomError : inOutRatio);
+
+                } else {
+                    Long inputBytes = snap.getStatusMetrics().get(INPUT_BYTES_METRIC);
+                    Long outputBytes = snap.getStatusMetrics().get(OUTPUT_BYTES_METRIC);
+                    Double inOutRatio = ((double) outputBytes / (double) inputBytes) + randomError;
+                    featureArray.add(Double.isNaN(inOutRatio) ? randomError : inOutRatio);
+                }
+
+                values.add((double) snapValue);
+                features.add(featureArray.toArray(new Double[featureArray.size()]));
+
+            }
+            return new Tuple<>(features.stream(), values.stream());
```

Comment:
If the values here will always be available as a `List`, does it make sense to convert to a Stream? Doing so ensure that any code that consumes this can only ever iterate over the values once. If it wants to iterate multiple times, it would have to convert into a List itself, which will eat up additional heap. Creating and iterating over the Stream is also less efficient than iterating over a List.